feat: add Alpha Vantage apdater + and support for API keys

nvtaveras opened the pull request:
## Description

This adds a first FX provider adapter (Alpha Vantage) which we will use for deploying the first XOF oracles on testnets. In addition to the adapter the client is also extended to support a new `API_KEYS` env variable in the format `<exchange-identifier-1>:<api-key-1>,<exchange-identifier-2>:<api-key-2>`, which is parsed and passed onto their respective adapters.

Other FX adapters will follow later on once we have more clarity on the final set of price sources.

## Other changes

- A bunch of linting changes done by prettier
- Updated the bitstamp certificate as I found out it was expired while testing (we currently don't use it but will soon for EUROC/EUR rates)

## Tested

- Wrote unit tests for the new adapter
- Ran the client locally using the new API keys env var

## Related issues

- Fixes https://github.com/mento-protocol/mento-general/issues/259

philbow61 commented on src/data_aggregator.ts:
```diff
@@ -1,42 +1,45 @@
 import { KrakenAdapter } from './exchange_adapters/kraken'
 import { KuCoinAdapter } from './exchange_adapters/kucoin'
-import { BitstampAdapter } from './exchange_adapters/bitstamp'
+import Logger from 'bunyan'
```
nit
maybe we can move this import to the top and not in between the aggregators

nvtaveras replied:
This was not me but the linter, so I will just end up moving it back

philbow61 commented on src/exchange_adapters/alphavantage.ts:
```diff
@@ -0,0 +1,142 @@
+    const base = AlphavantageAdapter.standardTokenSymbolMap.get(this.config.baseCurrency)
+    const quote = AlphavantageAdapter.standardTokenSymbolMap.get(this.config.quoteCurrency)
+
+    return `${base}${quote}`
```
wouldn't it make sense to return something like 
`from_currency=${base}&to_currency=${quote}`
and then use this.pairSymbol in fetchTicker and is OrderBookLive?

nvtaveras replied:
I see what you mean but I think it would be a bit strange since it's not really a symbol. For this adapter this function doesn't make a lot of sense but I had to implement it anyways

philbow61 replied:
ok makes sense

philbow61 commented on src/exchange_adapters/alphavantage.ts:
```diff
@@ -0,0 +1,142 @@
+    const ticker = this.parseTicker(tickerJson)
+
+    /*
+      TODO: verify the behaviour of the ticker endpoit on the weekend.
```
do we have an update on that?

nvtaveras replied:
Yeah I checked it on Sunday evening and the timestamp was still getting updated, which is not great. This is the response from Sunday 6th ~22:55 CET time:

```
{
    "Realtime Currency Exchange Rate": {
        "1. From_Currency Code": "XOF",
        "2. From_Currency Name": "CFA Franc BCEAO",
        "3. To_Currency Code": "EUR",
        "4. To_Currency Name": "Euro",
        "5. Exchange Rate": "0.00153000",
        "6. Last Refreshed": "2023-08-06 20:54:04",
        "7. Time Zone": "UTC",
        "8. Bid Price": "0.00152900",
        "9. Ask Price": "0.00153000"
    }
}
```

So based on that this check wouldn't work. However as we discussed yesterday, this is just for testing purposes so we'll roll with it for now and decide later on how do we want to approach this.

philbow61 reviewed (approved):
LGTM just some nits

nvtaveras commented on src/utils.ts:
```diff
@@ -81,6 +85,9 @@ export enum OracleCurrencyPair {
   EUROCUSDC = 'EUROCUSDC',
+  XOFEUR = 'XOFEUR',
+  USDXOF = 'USDXOF',
+  USDEUR = 'USDEUR',
```
removed!

bowd commented on src/exchange_adapters/alphavantage.ts:
```diff
@@ -0,0 +1,142 @@
+      timestamp: this.toUnixTimestamp(response[ResponseKeys.lastUpdated]),
+      // These FX API's do not provide volume data,
+      // therefore we set all of them to 1 to weight them equally
+      baseVolume: new BigNumber(1),
+      quoteVolume: new BigNumber(1),
```
How difficult would it be to have volume be optional here, and then assigning "fake" volume, somewhere else in the chain in a way that it won't get fucked if we mix "fake volume" exchange with "real volume" exchanges. 

Alternatively, could we add another field here, which is something like "volumeMissing: true" and set it to false in the `priceObjectMetadata` in `Base`, and then when aggregating throw an error unless either all of them are volumeMissing: true or false.

nvtaveras replied:
Updated this and left a comment [here](https://github.com/celo-org/celo-oracle/pull/182#issuecomment-1699157858) outlining the approach I took

bowd commented on src/exchange_adapters/alphavantage.ts:
```diff
@@ -0,0 +1,142 @@
+  }
+
+  toUnixTimestamp(datetime: string): number {
+    // datetime format: YYYY-MM-DD HH:MM:SS
+    const date = datetime.split(' ')[0]
+    const time = datetime.split(' ')[1]
+
+    const [year, month, day] = date.split('-').map(Number)
+    const [hours, minutes, seconds] = time.split(':').map(Number)
+
+    // month should be 0-indexed
+    return Date.UTC(year, month - 1, day, hours, minutes, seconds) / 1000
```
This is equivalent to `Date.parse(datetime) / 1000`

nvtaveras replied:
updated

bowd commented on src/exchange_adapters/alphavantage.ts:
```diff
@@ -0,0 +1,142 @@
+    const lastUpdated = ticker.timestamp // timestamp is already in seconds
+
+    assert(now >= lastUpdated, `Ticker timestamp is in the future: ${now} < ${lastUpdated}`)
+    return now - lastUpdated <= thirtyMinutesInSecs
```
If I understand correctly they don't actually return anything to indicate that the order book is closed, so we will manually pick a weekend range where we consider the order book not to be live.

nvtaveras replied:
the approach we will take is to find another source that returns stale data on the weekends and make the configuration n/n for the oracles to report. But yes, in this case as this exchange doesn't return anything I removed all of this code and just made this method return `true`

nvtaveras commented:
Updated with a bunch of changes since the last review, the most important one being 628ca4ee184ea33013b2da84fbb32ddf16314bd7 to add an `ignoreVolume` field to the price sources configuration. The change it's backwards compatible and the value will default to `false` when not passed in the configurations, so we don't need to update all the existing charts to add the extra field.

The idea is that we would set it for configurations that involve an FX rate along the way, for example, CELOXOF:

```
        {exchange: 'BINANCE', symbol: 'CELOUSDT', toInvert: false},
        {exchange: 'BINANCE', symbol: 'EURUSDT', toInvert: true},
        {exchange: 'ALPHAVANTAGE', symbol: 'EURXOF', toInvert: false, ignoreVolume: true}
```

while configurations that are only composed of FX rates (for example, EURXOF) don't need to set anything, as the default volume of `1` on FX adapters will ensure that they are all weighted equally.

The risk is that we could miss setting the property in a place where is actually necessary and end up "under-weighting" a price source, but imho this can just be part of the price source review process, similarly to how we need to ensure that the `toInvert` property is always correctly set.
